integration/daemon/daemon_test.go

```diff
@@ -381,6 +382,66 @@ func TestLiveRestore(t *testing.T) {
 	_ = testutil.StartSpan(baseContext, t)
 
 	t.Run("volume references", testLiveRestoreVolumeReferences)
+	t.Run("autoremove", testLiveRestoreAutoRemove)
+}
+
+func testLiveRestoreAutoRemove(t *testing.T) {
+	skip.If(t, testEnv.IsRootless(), "restarted rootless daemon will have a new process namespace")
+
+	t.Parallel()
+	ctx := testutil.StartSpan(baseContext, t)
+
+	run := func(t *testing.T) (*daemon.Daemon, func(), string) {
+		d := daemon.New(t)
+		d.StartWithBusybox(ctx, t, "--live-restore", "--iptables=false")
+		t.Cleanup(func() {
+			d.Stop(t)
+			d.Cleanup(t)
+		})
+
+		tmpDir := t.TempDir()
+
+		apiClient := d.NewClientT(t)
+
+		cID := container.Run(ctx, t, apiClient,
+			container.WithBind(tmpDir, "/v"),
+			// Run until a 'stop' file is created.
+			container.WithCmd("sh", "-c", "while [ ! -f /v/stop ]; do sleep 0.1; done"),
+			container.WithAutoRemove)
+		t.Cleanup(func() { apiClient.ContainerRemove(ctx, cID, containertypes.RemoveOptions{Force: true}) })
+		finishContainer := func() {
+			file, err := os.Create(filepath.Join(tmpDir, "stop"))
+			assert.NilError(t, err, "Failed to create 'stop' file")
+			file.Close()
+		}
+		return d, finishContainer, cID
+	}
+
+	t.Run("engine restart shouldnt kill alive containers", func(t *testing.T) {
+		d, finishContainer, cID := run(t)
+
+		d.Restart(t, "--live-restore", "--iptables=false")
+
+		apiClient := d.NewClientT(t)
+		_, err := apiClient.ContainerInspect(ctx, cID)
+		assert.NilError(t, err, "Container shouldn't be removed after engine restart")
+
+		finishContainer()
+
+		poll.WaitOn(t, container.IsRemoved(ctx, apiClient, cID))
+	})
+	t.Run("engine restart should remove containers that exited", func(t *testing.T) {
+		d, finishContainer, cID := run(t)
+
+		d.Stop(t)
+
+		finishContainer()
+		time.Sleep(time.Millisecond * 200)
```

Comment:
This is racey.
We could improve this by having the container write a file on exit and then wait 200ms, though that's still racey.

Another option would be to use a fifo, have the container open the fifo as a writer and the test open the file as read.
WHen the container exits the write side will be closed and read side will get EOF.

Reply:
Theoretically yes, in practice I think there's a very low probability that the container won't finish before the daemon is started.

With the FIFO I think there's still a race between closing the FIFO and the process being actually removed?

A simpler solution (and truly race-free I think?) would be to get the container process PID and just wait until it it's finished (`os.FindProcess` + `process.Wait()`). WDYT?

Reply:
> in practice I think there's a very low probability that the container won't finish before the daemon is started.

I've seen some stuff in our CI.
If it can go wrong it will.

> With the FIFO I think there's still a race between closing the FIFO and the process being actually removed?

Removed as in the shim seeing that it's down? Perhaps, but the fd will only be closed because the process exited.
Certainly some possible timing issues there in terms of what the shim would report, but I think that's the best we're going to get outside of making a separate request directly to the shim.

> A simpler solution (and truly race-free I think?) would be to get the container process PID and just wait until it it's finished (os.FindProcess + process.Wait()). WDYT?

I don't think you'll be able to wait for the process.
But... maybe in this case it will work since the test process would be the parent of dockerd->containerd->shim->process
Probably this has the same exact slight timing issue as above.
If you think this is simpler, that seems fine, though.

Reply:
Oh well, actually no.
This would use the main containerd process started with the test suite as I recall.

Reply:
If you want to wait on the pid, I think you'd need to first start an init container and then use that pid namespace for the test container.
This way the parent is the first container, as PID 1.

Reply:
Ah, right. I forgot that we can't `waitpid` on a non-child process...
There's still the `kill -0` trick (and we even seem to use it in `pkg/process`).

I did a quick implementation. Let me know if that looks good to you.